src/vision/status_quo/niklaus_lost_his_context.md

```diff
@@ -0,0 +1,99 @@
+
+At this point, Niklaus is looking at either rewriting a good chunk of the program to not use contexts, or rewriting his own code to sidestep the problem. He chooses the latter, by holding a reference to the thing that produces the context and then updating it after the async portion of the task has concluded:
+
+```use crate::{Index, Player}
+use wasm_bindgen_futures::spawn_local;
+use web_sys::window;
+use crate::parse_png;
+
+fn load_image(src: &url, inside_of: Index<Node>, player: Arc<Mutex<Player>>) {
+    spawn_local(async move {
+        let url = src.to_string();
+        let request = //still elided
+        let data = window().unwrap().fetch_with_request(&request).await.unwrap().etc.etc.etc;
+
+        player.lock().unwrap().update(|context| {
+            let image = parse_png(data, context);
+            context.display_tree.add_child_node(inside_of, image);
+        });
+    });
+}
+```
+
+It works, but it causes problems. Async and contextual code can't really coexist; so async code always has to live at the periphery of the program. More complicated async behaviors, such as reading the image in chunks and progressively decoding it, requires a delicate dance of await followed by update and so on. Niklaus tells Alan about this and he says it has something to do with colored functions, something he isn't quite familiar about.
+
+## 🤔 Frequently Asked Questions
+
+* **What are the morals of the story?**
+    * Borrowing means that Rust has three function colors, not two:
+      * Sync, which can both own and borrow it's parameters
+      * Owned Async, which can only own parameters
+      * Borrowed Async, which can both own and borrow parameters, but gains a lifetime
+    * Non-`'static` Futures are of limited use as lifetimes are tied to the sync stack. I think you can use `block_on` with them but that's about it.
+    * Borrowed Async functions would be useful if there was a way to have a future whose borrows were scoped to the moment they are `poll`ed.
+* **What are the sources for this story?**
+    * This is personal experience. Specifically, I had to do [almost exactly this dance][RuffleAsync] in order to get fetch to work in Ruffle.
+    * I have omitted a detail from this story: in Ruffle, we use a GC library (`gc_arena`) that imposes a special lifetime on all GC references. This is how the GC library upholds it's memory safety invariants, but it's also what forces us to pass around contexts, and once you have that, it's natural to start putting even non-GC data into it. It also means we can't hold anything from the GC in the Future as we cannot derive it's `Collect` trait on an anonymous type.
+* **Why did you choose Niklaus to tell this story?**
```

Comment:
I'm surprised by the choice of Niklaus since to me Niklaus is supposed to represent someone with little programming experience. The protagonist of this story seems to be fairly well versed in programming. They seem more like an Alan - familiar with JavaScript and gaining experience in Rust but not an expert.